src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -169,4 +277,48 @@ mod tests {
 
         assert_eq!(expected, mpc_sorted_list);
     }
+
+    #[tokio::test]
+    pub async fn malicious_sort_in_semi_honest() {
+        const COUNT: usize = 10;
+        const NUM_MULTI_BITS: u32 = 3;
+
+        let world = TestWorld::new().await;
+        let mut rng = thread_rng();
+
+        let mut match_keys = Vec::with_capacity(COUNT);
+        match_keys.resize_with(COUNT, || rng.gen::<MatchKey>());
+
+        let mut expected = match_keys.iter().map(|mk| mk.as_u128()).collect::<Vec<_>>();
+        expected.sort_unstable();
+
+        let [(v0, result0), (v1, result1), (v2, result2)] = world
+            .semi_honest(
+                match_keys.clone(),
+                |ctx: SemiHonestContext<Fp31>, mk_shares| async move {
+                    let local_lists = convert_all_bits_local(ctx.role(), &mk_shares);
+                    let converted_shares =
+                        convert_all_bits(&ctx, &local_lists, BitArray40::BITS, NUM_MULTI_BITS)
+                            .await
+                            .unwrap();
```

Comment:
This conversion wasn't validated... but it's a test for `malicious_generate_permutation_opt`, so I guess that doesn't really matter.

But in the actual IPA protocol we will need to validate this upgrade.

Reply:
The validate will be run before each reveal or downgrade of context right?